src/hidimstat/base_variable_importance.py

```diff
@@ -123,11 +143,182 @@ def selection(
+
+        if self.test_scores_.shape[0] == 1:
+            self.threshold_fdr_ = fdr_threshold(
+                self.pvalues_,
+                fdr=fdr,
+                method=fdr_control,
+                reshaping_function=reshaping_function,
+            )
+            selected = self.test_scores_[0] >= self.threshold_fdr_
+        elif not evalues:
+            assert fdr_control != "ebh", "for p-value, the fdr control can't be 'ebh'"
+            pvalues = np.array(
+                [_empirical_pval(test_score) for test_score in self.test_scores_]
+            )
+            self.aggregated_pval_ = quantile_aggregation(
+                pvalues, gamma=gamma, adaptive=adaptive_aggregation
+            )
+            self.threshold_fdr_ = fdr_threshold(
+                self.aggregated_pval_,
+                fdr=fdr,
+                method=fdr_control,
+                reshaping_function=reshaping_function,
             )
+            selected = self.aggregated_pval_ <= self.threshold_fdr_
+        else:
+            assert fdr_control == "ebh", "for e-value, the fdr control need to be 'ebh'"
+            evalues = []
+            for test_score in self.test_scores_:
+                ko_threshold = _estimated_threshold(test_score, fdr=fdr)
+                evalues.append(_empirical_eval(test_score, ko_threshold))
+            self.aggregated_eval_ = np.mean(evalues, axis=0)
+            self.threshold_fdr_ = fdr_threshold(
+                self.aggregated_eval_,
+                fdr=fdr,
+                method=fdr_control,
+                reshaping_function=reshaping_function,
+            )
+            selected = self.aggregated_eval_ >= self.threshold_fdr_
+        return selected
+
+
+def _estimated_threshold(test_score, fdr=0.1):
+    """
+    Calculate the threshold based on the procedure stated in the knockoff article.
+    Original code:
+    https://github.com/msesia/knockoff-filter/blob/master/R/knockoff/R/knockoff_filter.R
+    Parameters
+    ----------
+    test_score : 1D ndarray, shape (n_features, )
+        Vector of test statistic.
+    fdr : float
+        Desired controlled FDR (false discovery rate) level.
+    Returns
+    -------
+    threshold : float or np.inf
+        Threshold level.
+    """
+    offset = 1  # Offset equals 1 is the knockoff+ procedure.
+
+    threshold_mesh = np.sort(np.abs(test_score[test_score != 0]))
+    np.concatenate(
+        [[0], threshold_mesh, [np.inf]]
+    )  # if there is no solution, the threshold is inf
+    # find the right value of t for getting a good fdr
+    # Equation 1.8 of barber2015controlling and 3.10 in Candès 2018
+    threshold = 0.0
+    for threshold in threshold_mesh:
+        false_pos = np.sum(test_score <= -threshold)
+        selected = np.sum(test_score >= threshold)
+        if (offset + false_pos) / np.maximum(selected, 1) <= fdr:
+            break
+    return threshold
+
+
+def _empirical_pval(test_score):
```

Comment:
This function only makes sense in the context of knockoffs (it makes loyts of implicit assumptions on the so-called test_score) . I think it should be moved back to a knockoff module. If used outside of the knockoff context, it will return utterly wrong results.

_Originally posted by @bthirion in https://github.com/mind-inria/hidimstat/pull/359#discussion_r2311820110_

Reply:
Do we have a generic function for computing the threshold?

Are you sure that this can't be used also by Conditional Randomisation Tests and D0CRT?

Reply:
Yes, the standard procedure consists in the Benjamini-Hochberg algorithm. I believe that some past version of hidimstat included it. There is an implementation in Nilearn. It's essentially 2 lines of code.

Reply:
The Benjamini-Hockberg algorithm is implemented in the file multiple_testing.py and called through: `fdr_threshold`.

Reply:
I don't see the difference between CRT and knockoff from this point of view.
In principle, the method applied to knockoff can be applied to CRT because knockoff is just an approximation of CRT.

Reply:
I realise that the main difference between CRT and model_x knockoff is that model_x knockoff is doing multiple testing instead of doing independent testing as is done for CRT.